### ISSUE TITLE
DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.

### ISSUE THREAD

**Issue by its-me-ilia** [Automattic/mongoose#8156]
Hi, I just downloaded latest version of mongoose from npm and tried to run it but it gave me this warning:
DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.

i added this option to my configuration but it still gives me this warning. here is my configuration

`    mongoose.connect(process.env.MONGO_URI, {
        useNewUrlParser: true,
        useFindAndModify: false,
        useCreateIndex: true,
        useUnifiedTopology: true
    }).then(()=>{
        console.log(`connection to database established`)
    }).catch(err=>{
        console.log(`db error ${err.message}`);
        process.exit(-1)
    })`

any clue?

**Comment by felipeplets** [comment]
Same is happening to me. There is a thread about it also on StackOverflow: https://stackoverflow.com/questions/57895175/server-discovery-and-monitoring-engine-is-deprecated/57899638#57899638

**Comment by TREagans** [comment]
mongoose
  .connect(process.env.MONGO_URI, {
    useUnifiedTopology: true,
    useNewUrlParser: true,
  })
  .then(() => console.log('DB Connected!'))
  .catch(err => {
    console.log(`DB Connection Error: ${err.message}`);
  });

**Comment by portothree** [comment]
> This works fine for me, and no more errors.
> 
> mongoose
> .connect(process.env.MONGO_URI, {
> useUnifiedTopology: true,
> useNewUrlParser: true,
> })
> .then(() => console.log('DB Connected!'))
> .catch(err => {
> console.log(`DB Connection Error: ${err.message}`);
> });

I was facing the same error and "useUnifiedTopology: true" fixed, it's already updated on the readme of the repo btw

**Comment by jiobiagba** [comment]
> This works fine for me, and no more errors.
> 
> mongoose
> .connect(process.env.MONGO_URI, {
> useUnifiedTopology: true,
> useNewUrlParser: true,
> })
> .then(() => console.log('DB Connected!'))
> .catch(err => {
> console.log(`DB Connection Error: ${err.message}`);
> });

Adding the useUnifiedTopology: true to the object after my connection URL worked just as you suggested though I was using MongoClient. Thanks!

**Comment by kimprosh** [comment]
It still error:
```
const options = {
    keepAlive: 1,
    useUnifiedTopology: true,
    useNewUrlParser: true,
  };
  mongoose.connect(dbUrl, options).then(() => console.log('DB connected'));
```
It output:
```
(node:14923) DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.
```

**Comment by sinhagyandeep** [comment]
> > This works fine for me, and no more errors.
> > mongoose
> > .connect(process.env.MONGO_URI, {
> > useUnifiedTopology: true,
> > useNewUrlParser: true,
> > })
> > .then(() => console.log('DB Connected!'))
> > .catch(err => {
> > console.log(`DB Connection Error: ${err.message}`);
> > });
> 
> I was facing the same error and "useUnifiedTopology: true" fixed, thanks! It's already updated on the readme of the repo btw

@jiobiagba Yes i have tried your solution now my node.js server is not showing me any error simply  it is showing 'The Server Is Running On Port:7070'.

Thanks for the solution hope no error will come.

**Comment by douglasmartins7** [comment]
I was having the same problem and when inserting useUnifiedTopology: true, stopped showing terminal error.

Folow code:

mongo() {
    this.mongoConnection = mongoose.connect(process.env.MONGO_URL, {
      useUnifiedTopology: true,
      useNewUrlParser: true,
      useFindAndModify: true,
    });
  }

**Comment by cbratschi** [comment]
This parameter does not work with our replica set URL:

error:getaddrinfo ENOTFOUND .xyz.net .xyz.net:12345

Are there any replica set changes needed on top of this parameter?

**Comment by sinhagyandeep** [comment]
@cbratschi can you please explain in details cause i want exact error and code so further i can try to help you.

**Comment by guhandelta** [comment]
I too faced the same issue and after inserting useUnifiedTopology: true, I don't face any errors, but the console log in the .then() block does not get executed.

**Comment by sinhagyandeep** [comment]
@guhandelta none of code is executing (.then() and .catch()) ?

**Comment by guhandelta** [comment]
Yes... none, I don't get any console logs on my cmd prompt or errors.

server/index.js =>
![node indexjs](https://user-images.githubusercontent.com/17257834/66040812-6d5e0500-e4e6-11e9-9e80-149d257aab61.jpg)

npm run dev =>
![npm run dev](https://user-images.githubusercontent.com/17257834/66040813-6d5e0500-e4e6-11e9-8e04-5e517ea191a7.jpg)

**Comment by cbratschi** [comment]
> @cbratschi can you please explain in details cause i want exact error and code so further i can try to help you.

I only see these lines in the log message and the stacktrace contains the same information:

```
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:57:26)
(node:7740) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
```

This is the URL we are using (anonymized):

mongodb://user:pwd@abc.xyz.net:27018,.xyz.net:27019,.xyz.net:27020/account?replicaSet=mongors&readPreference=primaryPreferred

**Comment by sinhagyandeep** [comment]
@cbratschi can you please review your code once cause it will direct give the idea here the error is saying the same.it might possible that you are leaving some code for handling catch or async code or you might have missed  promise which is compulsory in non-blocking.

**Comment by pabloabreu1986** [comment]
Hello, I have:
```
mongoose
.connect(process.env.MONGO_URI, {
useUnifiedTopology: true,
useNewUrlParser: true,
})
.then(() => console.log('DB Connected!'))
.catch(err => {
console.log(Error, err.message);
});
```

When it starts I get at the console:
**`[Function: Error] { stackTraceLimit: 16, prepareStackTrace: undefined } connection 0 to acccluster-shard-00-01-xx47j.mongodb.net:27017 closed`**

**Comment by cbratschi** [comment]
> @cbratschi can you please review your code once cause it will direct give the idea here the error is saying the same.it might possible that you are leaving some code for handling catch or async code or you might have missed promise which is compulsory in non-blocking.

The subdomain was not parsed correctly in the implementation. Switching to full domains it works fine now (see my last comment). So the implementation is not fully standard compliant.

**Comment by franciscosuca** [comment]
> This works fine for me, and no more errors.
> 
> mongoose
> .connect(process.env.MONGO_URI, {
> useUnifiedTopology: true,
> useNewUrlParser: true,
> })
> .then(() => console.log('DB Connected!'))
> .catch(err => {
> console.log(`DB Connection Error: ${err.message}`);
> });

I was using the method indicated in this [link](https://mongoosejs.com/docs/deprecations.html), but it didn't work. After using the useUnifiedTopology option in the mongoose.connect() it worked!

**Comment by jbmulligan** [comment]
I found my way here because I am having the same problem - I just switched to the latest version of mongodb, and I get the warning.  I added useUnifiedTopology to the options passed to the MongoClient constructor, but the warning persists.  The connection to the remote database seems fine, the operations are succeeding and there are no other error messages or warnings.  My url is: mongodb://myUsername:myPassword@myHostName.org:27017/myDatabaseName
